File: src/main/java/ar/com/flow/bankaccount/Transfer.java
package ar.com.flow.bankaccount;

import ar.com.flow.bankaccount.transaction.Step;
import ar.com.flow.bankaccount.transaction.Transaction;
import ar.com.flow.bankaccount.transaction.receipt.Action;
import ar.com.flow.bankaccount.transaction.receipt.CreditPrinter;
import ar.com.flow.bankaccount.transaction.receipt.DebitPrinter;
import ar.com.flow.bankaccount.transfer.DifferentAccounts;
import ar.com.flow.bankaccount.withdrawal.SufficientFunds;
import ar.com.flow.money.Money;
import lombok.RequiredArgsConstructor;

class Transfer {
    static BankTransferBuilder from(BankAccount debitAccount) {
        return new BankTransferBuilder(debitAccount);
    }

    @RequiredArgsConstructor
    static class BankTransferBuilder {
        private final BankAccount debitAccount;
        private BankAccount creditAccount;

        BankTransferBuilder to(BankAccount creditAccount) {
            this.creditAccount = creditAccount;
            return this;
        }

        Transaction amount(Money amountToTransfer) {
            return Transaction.builder()
                    .precondition(new SufficientFunds(debitAccount, amountToTransfer))
                    .precondition(new DifferentAccounts(debitAccount, creditAccount))
                    .step(new Step(debitAccount, Transfer.debitReceipt(creditAccount)))
                    .step(new Step(creditAccount, Transfer.creditReceipt(creditAccount)))
                    .amount(amountToTransfer)
                    .build();
        }
    }

    static DebitPrinter debitReceipt(BankAccount account) {
        return new DebitPrinter(account, Action.Transfer);
    }

    static CreditPrinter creditReceipt(BankAccount account) {
        return new CreditPrinter(account, Action.Transfer);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/Transfer.kt
package ar.com.flow.bankaccount

import ar.com.flow.bankaccount.transaction.Builder
import ar.com.flow.bankaccount.transaction.steps.Step
import ar.com.flow.bankaccount.transaction.Transaction
import ar.com.flow.bankaccount.transaction.receipt.Action
import ar.com.flow.bankaccount.transaction.receipt.CreditPrinter
import ar.com.flow.bankaccount.transaction.receipt.DebitPrinter
import ar.com.flow.bankaccount.transfer.DifferentAccounts
import ar.com.flow.bankaccount.withdrawal.SufficientFunds
import ar.com.flow.money.Money

internal object Transfer {
    fun from(debitAccount: BankAccount): BankTransferBuilder {
        return BankTransferBuilder(debitAccount)
    }

    fun debitReceipt(account: BankAccount): DebitPrinter {
        return DebitPrinter(account, Action.Transfer)
    }

    fun creditReceipt(account: BankAccount): CreditPrinter {
        return CreditPrinter(account, Action.Transfer)
    }

    internal class BankTransferBuilder(
        private val debitAccount: BankAccount
    ) {
        private lateinit var creditAccount: BankAccount

        fun to(creditAccount: BankAccount): BankTransferBuilder {
            this.creditAccount = creditAccount
            return this
        }

        fun amount(amountToTransfer: Money): Transaction {
            return Builder()
                .precondition(
                    SufficientFunds(
                        debitAccount,
                        amountToTransfer
                    )
                )
                .precondition(
                    DifferentAccounts(
                        debitAccount,
                        creditAccount
                    )
                )
                .step(Step(debitAccount, debitReceipt(creditAccount)))
                .step(Step(creditAccount, creditReceipt(creditAccount)))
                .amount(amountToTransfer)
                .build()
        }
    }
}